packages/federation/src/managed-federation.ts

```diff
@@ -0,0 +1,193 @@
+   */
+  lastSeenId?: string;
+  /**
+   * The fetch implementation to use.
+   * Default: global.fetch
+   */
+  fetch?: FetchFn;
+};
+
+export type RouterConfig = {
+  supergraphSdl: string;
+  minDelaySeconds: number;
+  id: string;
+};
+
+export type Unchanged = {
+  minDelaySeconds?: never;
+  id: string;
+};
+
+export type FetchError = {
+  error: { code: string; message: string };
+  minDelaySeconds: never;
+};
+
+type RouterConfigResult = {
+  routerConfig:
+    | ({ __typename: 'RouterConfigResult' } & RouterConfig)
+    | ({ __typename: 'Unchanged' } & Unchanged)
+    | { __typename: 'FetchError'; code: string; message: string; minDelaySeconds: never };
+};
+
+export const DEFAULT_UPLINK_ENDPOINTS = [
+  'https://uplink.api.apollographql.com/',
+  'https://aws.uplink.api.apollographql.com/',
+];
+
+/**
+ * Fetches the supergraph SDL from a managed federation GraphOS up link.
+ * @param options
+ * @returns An object with the supergraph SDL when possible. It also includes metadata to handle polling and retry logic.
+ *          If `lastSeenId` is provided and the supergraph has not changed, `supergraphSdl` is not present.
+ */
+export async function fetchSupergraphSdlFromManagedFederation(
+  options: FetchSupergraphSdlFromManagedFederationOpts,
+): Promise<RouterConfig | Unchanged | FetchError> {
+  const { upLink = DEFAULT_UPLINK_ENDPOINTS[0], fetch = defaultFetch, ...variables } = options;
+
+  const response = await fetch(upLink, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({
+      query: /* GraphQL */ `
+        query ($apiKey: String!, $graphId: String!, $lastSeenId: ID) {
+          routerConfig(ref: $graphId, apiKey: $apiKey, ifAfterId: $lastSeenId) {
+            __typename
+            ... on FetchError {
+              code
+              message
+              minDelaySeconds
+            }
+            ... on Unchanged {
+              id
+              minDelaySeconds
+            }
+            ... on RouterConfigResult {
+              id
+              supergraphSdl: supergraphSDL
+              minDelaySeconds
+            }
+          }
+        }
+      `,
+      variables,
+    }),
+  });
+
+  const responseBody = await response.text();
+
+  if (!response.ok) {
+    throw new Error(
+      `Failed to fetch supergraph SDL from managed federation up link '${upLink}': [${response.status} ${response.statusText}] ${responseBody}`,
+    );
+  }
+
+  let result: ExecutionResult<RouterConfigResult>;
+  try {
+    result = JSON.parse(responseBody);
+  } catch (err) {
+    throw new Error(
+      `Failed to parse response from managed federation up link '${upLink}': ${(err as Error).message}\n\n${responseBody}`,
+    );
+  }
+
+  if (!result.data?.routerConfig) {
+    throw new Error(
+      `Failed to fetch supergraph SDL from managed federation up link '${upLink}': ${responseBody}`,
+    );
+  }
```

Comment:
This is not necessary. the `routerConfig` field is mandatory and not nullable, which means it has to be there, or `errors` will be present. This will disallow to actually show the errors in the next condition.

Reply:
But data is optional